Classes/Domain/NodeCreation/NodeCreationElements.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neos\Neos\Ui\Domain\NodeCreation;
+
+use Neos\ContentRepository\Core\SharedModel\Node\NodeAggregateIds;
+
+/**
+ * Access to deserialize elements of the node creation dialog
+ *
+ * Property-like elements are of simple types or objects.
+ * The values will be deserialized according to its type.
+ * For example myImage will be an actual image object instance.
+ *
+ *     Vendor.Site:Content:
+ *       ui:
+ *         creationDialog:
+ *           elements:
+ *             myString:
+ *               type: string
+ *             myImage:
+ *               type: Neos\Media\Domain\Model\ImageInterface
+ *
+ * Reference-like elements are of type `references` or `reference`
+ * And will be available as NodeAggregateIds collection.
+ *
+ *     Vendor.Site:Content:
+ *       ui:
+ *         creationDialog:
+ *           elements:
+ *             myReferences:
+ *               type: references
+ *
+ * The same categories apply to promoted elements:
+ *
+ *     Vendor.Site:Content:
+ *       properties:
+ *         myString:
+ *           type: string
+ *           ui:
+ *             showInCreationDialog: true
+ *
+ * @api As part of the {@see NodeCreationHandlerInterface}
+ */
+final readonly class NodeCreationElements
+{
+    /**
+     * @param array<string, mixed> $propertyLikeValues
+     * @param array<string, NodeAggregateIds> $referenceLikeValues
+     * @param array<int|string, mixed> $serializedValues
+     * @internal you should not need to construct this
+     */
+    public function __construct(
+        private array $propertyLikeValues,
+        private array $referenceLikeValues,
```

Comment:
```suggestion
        private array $elementValues,
```

array<string, NodeAggregateIds|mixed>

Reply:
as far as I understand it, this DTO is an object that can be used to prime the creation of nodes with some parameters and we (currently) use it to map the values of the NodeCreation dialog.

As mentioned in our weekly today, I wonder why we need to separate the values already at this point. If we _do_ I would suggest `$propertyValues` and `$referencedNodeAggregateIds` but I would personally prefer to keep this as a simple key-value DTO for arbitrary parameters.

That would mean that the code below would instead of
```php
if (!$elements->hasReferenceLike($referenceName)) {
    continue;
}
$setReferencesCommands[] = SetNodeReferences::create(
    $commands->first->contentStreamId,
    $commands->first->nodeAggregateId,
    $commands->first->originDimensionSpacePoint,
      ReferenceName::fromString($referenceName),
      NodeReferencesToWrite::fromNodeAggregateIds($elements->getReferenceLike($referenceName))
);
```
be s.th. like:
```php
if (!$parameters->has($referenceName)) {
    continue;
}

$setReferencesCommands[] = SetNodeReferences::create(
    $commands->first->contentStreamId,
    $commands->first->nodeAggregateId,
    $commands->first->originDimensionSpacePoint,
      ReferenceName::fromString($referenceName),
      NodeReferencesToWrite::fromNodeAggregateIds(NodeAggregateIds::fromArrray($parameters->get($referenceName)),
);
```

(potentially with some error handling if the expected format does not match)

And I would even leave out the deserialization at this point..